File: code/client/munkilib/installer/pkg.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
installer.pkg

Created by Greg Neagle on 2017-01-03.

Routines for installing Apple pkgs
"""

# This code is largely still compatible with Python 2, so for now, turn off
# Python 3 style warnings
# pylint: disable=consider-using-f-string
# pylint: disable=redundant-u-string-prefix

from __future__ import absolute_import, print_function

import os
import pwd
import signal
import subprocess
import time

from .. import display
from .. import dmgutils
from .. import info
from .. import munkilog
from .. import munkistatus
from .. import osutils
from .. import processes
from .. import pkgutils
from .. import FoundationPlist


def remove_bundle_relocation_info(pkgpath):
    '''Attempts to remove any info in the package that would cause
    bundle relocation behavior. This makes bundles install or update in their
    default location.'''
    display.display_debug1("Looking for bundle relocation info...")
    if os.path.isdir(pkgpath):
        # remove relocatable stuff
        tokendefinitions = os.path.join(
            pkgpath, "Contents/Resources/TokenDefinitions.plist")
        if os.path.exists(tokendefinitions):
            try:
                os.remove(tokendefinitions)
                display.display_debug1(
                    "Removed Contents/Resources/TokenDefinitions.plist")
            except OSError:
                pass

        plist = {}
        infoplist = os.path.join(pkgpath, "Contents/Info.plist")
        if os.path.exists(infoplist):
            try:
                plist = FoundationPlist.readPlist(infoplist)
            except FoundationPlist.NSPropertyListSerializationException:
                pass

        if 'IFPkgPathMappings' in plist:
            del plist['IFPkgPathMappings']
            try:
                FoundationPlist.writePlist(plist, infoplist)
                display.display_debug1("Removed IFPkgPathMappings")
            except FoundationPlist.NSPropertyListWriteException:
                pass


def pkg_needs_restart(pkgpath, options):
    '''Query a package for its RestartAction. Returns True if a restart is
    needed, False otherwise'''
    cmd = ['/usr/sbin/installer', '-query', 'RestartAction', '-pkg', pkgpath]
    if options.get('installer_choices_xml'):
        choices_xml_file = os.path.join(osutils.tmpdir(), 'choices.xml')
        FoundationPlist.writePlist(
            options.get('installer_choices_xml'), choices_xml_file)
        cmd.extend(['-applyChoiceChangesXML', choices_xml_file])
    else:
        choices_xml_file = None
    if options.get('allow_untrusted'):
        cmd.append('-allowUntrusted')
    proc = subprocess.Popen(cmd, shell=False, bufsize=-1,
                            stdin=subprocess.PIPE,
                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    output = proc.communicate()[0].decode('UTF-8')
    restartaction = output.rstrip('\n')
    return restartaction in ['RequireRestart', 'RecommendRestart']


def get_installer_env(custom_env):
    '''Sets up environment for installer'''
    env_vars = os.environ.copy()
    # get info for root
    userinfo = pwd.getpwuid(0)
    env_vars['USER'] = userinfo.pw_name
    env_vars['HOME'] = userinfo.pw_dir
    if custom_env:
        # Munki admin has specified custom installer environment
        for key in custom_env.keys():
            if key == 'USER' and custom_env[key] == 'CURRENT_CONSOLE_USER':
                # current console user (if there is one) 'owns' /dev/console
                userinfo = pwd.getpwuid(os.stat('/dev/console').st_uid)
                env_vars['USER'] = userinfo.pw_name
                env_vars['HOME'] = userinfo.pw_dir
            else:
                env_vars[key] = custom_env[key]
        display.display_debug1(
            'Using custom installer environment variables: %s', env_vars)
    return env_vars


def _display_installer_output(installinfo):
    '''Parses a line of output from installer, displays it as progress output
    and logs it'''
    # output we're dealing with always starts with 'installer:'
    msg = installinfo[10:].rstrip("\n")
    if msg.startswith("PHASE:"):
        phase = msg[6:]
        if phase:
            display.display_status_minor(phase)
    elif msg.startswith("STATUS:"):
        status = msg[7:]
        if status:
            display.display_status_minor(status)
    elif msg.startswith("%"):
        percent = float(msg[1:])
        munkistatus.percent(percent)
        display.display_status_minor("%s percent complete" % percent)
    elif msg.startswith(" Error"):
        display.display_error(msg)
        munkistatus.detail(msg)
    elif msg.startswith(" Cannot install"):
        display.display_error(msg)
        munkistatus.detail(msg)
    else:
        munkilog.log(msg)


def _run_installer(cmd, env_vars, packagename):
    '''Runs /usr/sbin/installer, parses and displays the output, and returns
    the process exit code'''
    installeroutput = []

    job = subprocess.Popen(
        cmd,
        env=env_vars,
        shell=False,
        bufsize=-1,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    timeout = 2 * 60 * 60
    last_output = None
    while True:
        installinfo = job.stdout.readline()
        if not installinfo and job.poll() is not None:
            break

        # Don't bother parsing the stdout output if it hasn't changed since
        # the last loop iteration.
        if last_output == installinfo:
            continue
        last_output = installinfo

        installinfo = installinfo.decode('UTF-8')
        if installinfo.startswith("installer:"):
            # save all installer output in case there is
            # an error so we can dump it to the log
            installeroutput.append(installinfo)
            _display_installer_output(installinfo)

    # installer exited or was killed
    # try for a little bit to catch return code from exiting process...
    retcode = job.poll()
    i = 0
    while retcode is None and i < 5:
        time.sleep(1)
        i += 1
        retcode = job.poll()

    if retcode != 0:
        # append stdout to our installer output
        installeroutput.extend(job.stderr.read().decode("UTF-8").splitlines())
        display.display_status_minor(
            "Install of %s failed with return code %s" % (packagename, retcode))
        display.display_error("-"*78)
        for line in installeroutput:
            display.display_error(line.rstrip("\n"))
        display.display_error("-"*78)
    elif retcode == 0:
        munkilog.log("Install of %s was successful." % packagename)
        munkistatus.percent(100)
    return retcode


def install(pkgpath, options=None):
    """
    Uses the apple installer to install the package or metapackage
    at pkgpath.
    Returns a tuple:
    the installer return code and restart needed as a boolean.
    """

    restartneeded = False

    if not options:
        options = {}
    display_name = options.get('display_name') or options.get('name')

    if os.path.islink(pkgpath):
        # resolve links before passing them to /usr/bin/installer
        pkgpath = os.path.realpath(pkgpath)

    if options.get('suppress_bundle_relocation'):
        remove_bundle_relocation_info(pkgpath)

    packagename = os.path.basename(pkgpath)
    if not display_name:
        display_name = packagename

    munkilog.log("Installing %s from %s" % (display_name, packagename))
    if pkg_needs_restart(pkgpath, options):
        display.display_status_minor(
            '%s requires a restart after installation.' % display_name)
        restartneeded = True

    # set up installer cmd
    cmd = ['/usr/sbin/installer', '-verboseR', '-pkg', pkgpath, '-target', '/']
    if options.get('installer_choices_xml'):
        # choices_xml_file was already built by pkg_needs_restart(),
        # just re-use it
        choices_xml_file = os.path.join(osutils.tmpdir(), 'choices.xml')
        cmd.extend(['-applyChoiceChangesXML', choices_xml_file])
    if options.get('allow_untrusted'):
        cmd.append('-allowUntrusted')

    # get env for installer
    env_vars = get_installer_env(options.get('installer_environment'))

    # run it!
    retcode = _run_installer(cmd, env_vars, packagename)
    if retcode:
        restartneeded = False
    return (retcode, restartneeded)


def installall(dirpath, options=None):
    """
    Attempts to install all pkgs and mpkgs in a given directory.
    Will mount dmg files and install pkgs and mpkgs found at the
    root of any mountpoints.
    """
    retcode = 0
    restartflag = False
    installitems = osutils.listdir(dirpath)
    for item in installitems:
        if processes.stop_requested():
            return (retcode, restartflag)
        itempath = os.path.join(dirpath, item)
        if pkgutils.hasValidDiskImageExt(item):
            display.display_info("Mounting disk image %s" % item)
            mountpoints = dmgutils.mountdmg(
                itempath, use_shadow=True, skip_verification=True)
            if not mountpoints:
                display.display_error("No filesystems mounted from %s", item)
                return (retcode, restartflag)
            if processes.stop_requested():
                dmgutils.unmountdmg(mountpoints[0])
                return (retcode, restartflag)
            for mountpoint in mountpoints:
                # install all the pkgs and mpkgs at the root
                # of the mountpoint -- call us recursively!
                (retcode, needsrestart) = installall(
                    mountpoint, options=options)
                if needsrestart:
                    restartflag = True
                if retcode:
                    # ran into error; should unmount and stop.
                    dmgutils.unmountdmg(mountpoints[0])
                    return (retcode, restartflag)

            dmgutils.unmountdmg(mountpoints[0])

        if pkgutils.hasValidInstallerItemExt(item):
            (retcode, needsrestart) = install(
                itempath, options=options)
            if needsrestart:
                restartflag = True
            if retcode:
                # ran into error; should stop.
                return (retcode, restartflag)

    return (retcode, restartflag)


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/installer/pkg.swift
//
//  pkg.swift
//  munki
//
//  Created by Greg Neagle on 8/3/24.
//
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

private let display = DisplayAndLog.main

/// Stub function
func removeBundleRelocationInfo() {
    display.warning("'suppress_bundle_relocation' is no longer supported. Ignoring.")
}

/// Query a package for its RestartAction. Returns true if a restart is needed, false otherwise
func pkgNeedsRestart(_ pkgpath: String, options: PlistDict) -> Bool {
    let tool = "/usr/sbin/installer"
    var arguments = ["-query", "RestartAction", "-pkg", pkgpath, "-plist"]
    if let choicesXML = options["installer_choices_xml"] as? PlistDict,
       let tempDir = TempDir.shared.path
    {
        let choicesXMLPath = (tempDir as NSString).appendingPathComponent("choices.xml")
        do {
            try writePlist(choicesXML, toFile: choicesXMLPath)
            arguments += ["-applyChoiceChangesXML", choicesXMLPath]
        } catch {}
    }
    if let allowUntrusted = options["allow_untrusted"] as? Bool,
       allowUntrusted == true
    {
        arguments.append("-allowUntrusted")
    }
    let result = runCLI(tool, arguments: arguments)
    if result.exitcode != 0 {
        display.warning("/usr/bin/installer error when getting restart info for \((pkgpath as NSString).lastPathComponent): \(result.error)")
        return false
    }
    let (pliststr, _) = parseFirstPlist(fromString: result.output)
    if !pliststr.isEmpty,
       let plist = try? readPlist(fromString: pliststr) as? PlistDict,
       let restartAction = plist["RestartAction"] as? String
    {
        return ["RequireRestart", "RecommendRestart"].contains(restartAction)
    }
    display.warning("/usr/bin/installer returned unexpected value when getting restart info for \((pkgpath as NSString).lastPathComponent): \(result.output)")
    return false
}

func getInstallerEnvironment(_ customEnv: [String: String]?) -> [String: String] {
    var env = ProcessInfo.processInfo.environment
    env["USER"] = NSUserName()
    env["HOME"] = NSHomeDirectory()
    if let customEnv {
        for (key, value) in customEnv {
            if key == "USER", value == "CURRENT_CONSOLE_USER" {
                var consoleUser = getConsoleUser()
                if consoleUser == "" || consoleUser == "loginwindow" {
                    consoleUser = "root"
                }
                env["USER"] = consoleUser
                env["HOME"] = NSHomeDirectoryForUser(consoleUser)
            } else {
                env[key] = value
            }
        }
        display.debug1("Using custom installer environment variables: \(env)")
    }
    return env
}

/// Parses a line of output from installer, displays it as progress output and logs it
func displayInstallerOutput(_ text: String) {
    if !text.hasPrefix("installer:") {
        // this should not have been sent to this function!
        return
    }
    var msg = text.trimmingCharacters(in: .newlines)
    // delete "installer:" prefix
    msg.removeFirst("installer:".count)
    if msg.hasPrefix("PHASE:") {
        msg.removeFirst("PHASE:".count)
        if !msg.isEmpty {
            display.minorStatus(msg)
        }
    } else if msg.hasPrefix("STATUS:") {
        msg.removeFirst("STATUS:".count)
        if !msg.isEmpty {
            display.minorStatus(msg)
        }
    } else if msg.hasPrefix("%") {
        msg.removeFirst()
        if let percent = Double(msg) {
            munkiStatusPercent(Int(percent))
            display.minorStatus("\(msg) percent complete")
        }
    } else if msg.hasPrefix(" Error") || msg.hasPrefix(" Cannot install") {
        display.error(msg)
        munkiStatusDetail(msg)
    } else {
        munkiLog(msg)
    }
}

/// Subclass of AsyncProcessRunner that handles the progress output from /usr/sbin/installer
class installerRunner: AsyncProcessRunner {
    var remainingOutput = ""
    var lastProcessedOutputLine = ""

    override init(_ tool: String = "/usr/sbin/installer",
                  arguments: [String] = [],
                  environment: [String: String] = [:],
                  stdIn: String = "")
    {
        super.init(tool, arguments: arguments, environment: environment, stdIn: stdIn)
    }

    func linesAndRemainderOf(_ str: String) -> ([String], String) {
        var lines = str.components(separatedBy: "\n")
        var remainder = ""
        if lines.count > 0, !str.hasSuffix("\n") {
            // last line of string did not end with a newline; might be a partial
            remainder = lines.last ?? ""
            lines.removeLast()
        }
        return (lines, remainder)
    }

    override func processOutput(_ str: String) {
        super.processOutput(str)
        let (lines, remainder) = linesAndRemainderOf(remainingOutput + str)
        remainingOutput = remainder
        for line in lines {
            if line == lastProcessedOutputLine {
                // if the output line is the same, just skip
                continue
            } else if line.hasPrefix("installer:") {
                displayInstallerOutput(line)
                lastProcessedOutputLine = line
            }
        }
    }
}

/// Runs /usr/sbin/installer, parses and displays the output, and returns the process exit code
func runInstaller(arguments: [String], environment: [String: String], pkgName: String) async -> Int {
    let proc = installerRunner(arguments: arguments, environment: environment)
    await proc.run()
    let results = proc.results
    if results.exitcode != 0 {
        display.minorStatus("Install of \(pkgName) failed with return code \(results.exitcode)")
        display.error(String(repeating: "-", count: 78))
        for line in results.output.components(separatedBy: "\n") {
            display.error(line)
        }
        for line in results.error.components(separatedBy: "\n") {
            display.error(line)
        }
        display.error(String(repeating: "-", count: 78))
    }
    return results.exitcode
}

// Uses the Apple installer to install the package or metapackage at pkgpath.
// Returns a tuple:
//    the installer return code and restart needed as a boolean.
func install(_ pkgpath: String, options: PlistDict = [:]) async -> (Int, Bool) {
    var restartNeeded = false
    let packageName = (pkgpath as NSString).lastPathComponent
    let displayName = options["display_name"] as? String ?? options["name"] as? String ?? packageName

    var resolvedPkgPath = pkgpath
    if pathIsSymlink(pkgpath) {
        resolvedPkgPath = getAbsolutePath(pkgpath)
    }

    if let suppressBundleRelocation = options["suppress_bundle_relocation"] as? Bool, suppressBundleRelocation == true {
        removeBundleRelocationInfo()
    }

    munkiLog("Installing \(displayName) from \(packageName)")
    if pkgNeedsRestart(resolvedPkgPath, options: options) {
        display.minorStatus("\(displayName) requires a restart after installation.")
        restartNeeded = true
    }

    var arguments = ["-verboseR", "-pkg", resolvedPkgPath, "-target", "/"]
    if let choicesXML = options["installer_choices_xml"] as? [PlistDict],
       let tempDir = TempDir.shared.path
    {
        let choicesXMLPath = (tempDir as NSString).appendingPathComponent("choices.xml")
        do {
            try writePlist(choicesXML, toFile: choicesXMLPath)
            arguments += ["-applyChoiceChangesXML", choicesXMLPath]
        } catch {
            // could not write choices.xml, should not proceed
            display.error("Could not write choices.xml for \(packageName)")
            return (-1, false)
        }
    }
    if let allowUntrusted = options["allow_untrusted"] as? Bool,
       allowUntrusted == true
    {
        arguments.append("-allowUntrusted")
    }

    // get installer environment
    let envVars = getInstallerEnvironment(options["installer_environment"] as? [String: String])

    // run it
    let retcode = await runInstaller(arguments: arguments, environment: envVars, pkgName: packageName)
    if retcode != 0 {
        restartNeeded = false
    }
    return (retcode, restartNeeded)
}

// The Python version of Munki would actually install _all_ the pkgs from a given
// directory (which was usually the root of a mounted disk image). This was rarely
// what was actually wanted. This version just installs the first installable item in
// the directory.
// Returns a tuple containing the exit code of the installer process and a boolean
// indicating if a restart is needed
func installFromDirectory(_ directoryPath: String, options: PlistDict = [:]) async -> (Int, Bool) {
    if stopRequested() {
        return (0, false)
    }
    if let items = try? FileManager.default.contentsOfDirectory(atPath: directoryPath) {
        for item in items {
            let itempath = (directoryPath as NSString).appendingPathComponent(item)
            if hasValidDiskImageExt(item) {
                display.info("Mounting disk image \(item)")
                guard let mountpoint = try? mountdmg(itempath, useShadow: true, skipVerification: true) else {
                    display.error("No filesystems mounted from \(item)")
                    return (-1, false)
                }
                // make sure we unmount this when done
                defer {
                    do {
                        try unmountdmg(mountpoint)
                    } catch {
                        display.error(error.localizedDescription)
                    }
                }
                // call us recursively to install a pkg at the root of this diskimage
                return await installFromDirectory(mountpoint, options: options)
            }
            if hasValidPackageExt(item) {
                return await install(itempath, options: options)
            }
        }
    }
    // if we get here, no valid items to install were found
    display.warning("No items to install were found in \(directoryPath)")
    return (-1, false)
}
